tests/test_bqpipeline.py

```diff
@@ -63,24 +77,116 @@ def test_create_job_config_destination(self):
+        self.assertEqual(result, bigquery.ScalarQueryParameter('test',
+                                                               'STRING', 'abc'))
+
+        result = bqpipeline.set_scalar_parameter('test', 1, typ='int')
+        self.assertEqual(result, bigquery.ScalarQueryParameter('test',
+                                                               'INT64', 1))
+
+        dt = datetime.datetime.now()
+        result = bqpipeline.set_scalar_parameter('test', dt, typ='datetime')
+        self.assertEqual(result, bigquery.ScalarQueryParameter('test',
+                                                               'TIMESTAMP', dt))
+
+        b = bytes(123)
+        result = bqpipeline.set_scalar_parameter('test', b, typ='bytes')
+        self.assertEqual(result, bigquery.ScalarQueryParameter('test',
+                                                               'BYTES', b))
+
+        with self.assertRaises(KeyError):
+            bqpipeline.set_scalar_parameter('test', 1, typ='unsupported')
+
+    def test_array_parameters(self):
+        result = bqpipeline.set_list_parameter('test', ['abc'], typ='list')
+        self.assertEqual(result, bigquery.ArrayQueryParameter('test', 'STRING',
+                                                              ['abc']))
+
+        result = bqpipeline.set_list_parameter('test', [1, 2], typ='list')
+        self.assertEqual(result, bigquery.ArrayQueryParameter('test', 'INT64',
+                                                              [1, 2]))
+
+        with self.assertRaises(IndexError):
+            bqpipeline.set_list_parameter('test', [], typ='list')
+
+    def test_struct_parameters(self):
+        result = bqpipeline.set_dict_parameter('test', {'a': 'abc'},
+                                               typ='list')
+        self.assertEqual(result, bigquery.StructQueryParameter(
+            'test',
+            bigquery.ScalarQueryParameter('a', 'STRING', 'abc')))
+
+    def test_positional_parameters(self):
+        result = bqpipeline.set_scalar_parameter(None, 'abc', typ='str')
+        self.assertEqual(result, bigquery.ScalarQueryParameter(None,
+                                                               'STRING', 'abc'))
+
+        result = bqpipeline.set_scalar_parameter(None, 1, typ='int')
+        self.assertEqual(result, bigquery.ScalarQueryParameter(None,
+                                                               'INT64', 1))
+
+        result = bqpipeline.set_list_parameter(None, ['abc'], typ='list')
+        self.assertEqual(result, bigquery.ArrayQueryParameter(None, 'STRING',
+                                                              ['abc']))
+
+    def test_query_params(self):
+        bq = bqpipeline.BQPipeline(
+            job_name='testjob', default_project='testproject',
+            default_dataset='testdataset')
+        cfg = bq.create_job_config(batch=False, create=False, overwrite=False)
+        self.assertIsNone(bq.set_query_params([]))
```

Comment:
See note above on having `create_job_config` accept a `query_params` arg.
We should refactor this behavior and then have a test of `create_job_config(...,query_params)` and validate it creates the whole job_config as expected for no query params and some query params.

Reply:
Done. Also, note the change in method invocation. Named keyword parameters should always be passed as such otherwise, they may get wrong value unintentionally.